Fixes #25439 - Add new DiffView to Config Reports

glekner opened the pull request:
Config Reports should now show diffs with the new DiffView Component.

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/24938324/48780410-7f75f100-ece2-11e8-8720-687664ade248.gif)

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [Translating section in the guide]
(https://projects.theforeman.org/projects/foreman/wiki/Translating)
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* Be nice and respectful

We are running bots that will poke you if you miss an item from the list :-)

--->

glekner committed:
fixes #25439 - Add new DiffView to Config Reports

ohadlevy commented:
@glekner can you please include a short screenshot/video of the feature? thanks.

ohadlevy commented on app/assets/javascripts/reports.js:
```diff
@@ -1,26 +0,0 @@
-$(function() {
```
please remove the reference to the reports.js, .e.g. at https://github.com/theforeman/foreman/blob/d7f8cdd651cf1060bf28abd93dbceac8e72eccd1/app/views/config_reports/show.html.erb#L1

(maybe we can remove the charts js as well if we migrated away to react charts?)

ohadlevy commented on app/helpers/reports_helper.rb:
```diff
@@ -67,4 +67,11 @@ def report_origin_output_partial(origin)
     'output'.freeze
   end
+
+  def format_message(message, log)
```
maybe give this a better name? something like config_report_content ?

ohadlevy commented on app/views/config_reports/show.html.erb:
```diff
@@ -20,20 +20,6 @@
   link_to(_("Host details"), @config_report.host, :class => 'btn btn-default'),
   link_to(_("Other reports for this host"), host_config_reports_path(@config_report.host), :class => 'btn btn-default')
 %>
+  <div  id="diff-modal-editor" class="modal-body"></div>
```
nitpick: extra whitespaces?

ohadlevy commented on webpack/assets/javascripts/bundle.js:
```diff
@@ -42,4 +42,5 @@ window.tfm = Object.assign(window.tfm || {}, {
   advancedFields: require('./foreman_advanced_fields'),
   i18n,
   breadcrumbs: require('./foreman_breadcrumbs'),
+  diffModal: require('./foreman_modal_diff'),
```
should this be in the context of config reports? or rather you see this component for everywhere you would like to show a diff in a modal?

ohadlevy commented on webpack/assets/javascripts/react_app/components/AuditsList/__tests__/__snapshots__/ExpansiveView.test.js.snap:
```diff
@@ -7,6 +7,7 @@ exports[`ExpansiveView rendering render audit record changes 1`] = `
   <DiffContainer
     newText="<h1>Hello World..</h1>"
     oldText="<h1>Hello..</h1>"
+    patch=""
```
why?

glekner replied:
beacuse the diffView now supports both old,new Text and patch inputs. only one of the options is used at a time

ohadlevy commented on webpack/assets/javascripts/react_app/components/ConfigReports/DiffModal/DiffModal.js:
```diff
@@ -0,0 +1,48 @@
+    return (
+      <Modal show={isOpen} onHide={() => toggleModal()} className="diff-modal">
+        <Modal.Header>
+          <h4 id="diff-modal-h4">{log}</h4>
```
whats log? is it a title?

ohadlevy commented on webpack/assets/javascripts/react_app/components/ConfigReports/DiffModal/DiffModal.js:
```diff
@@ -0,0 +1,48 @@
+      <Modal show={isOpen} onHide={() => toggleModal()} className="diff-modal">
+        <Modal.Header>
+          <h4 id="diff-modal-h4">{log}</h4>
+          <button
```
use PF Button component?

ohadlevy commented on webpack/assets/javascripts/react_app/components/ConfigReports/DiffModal/DiffModal.js:
```diff
@@ -0,0 +1,48 @@
+DiffModal.defaultProps = {
+  patch: '',
+  oldText: '',
+  newText: '',
```
should we allow to open it if there is no diff? (Does it even work?)

glekner replied:
The component won't even render because the parent checks if there is a diff. (Audits and Here).

ohadlevy commented on webpack/assets/javascripts/react_app/components/ConfigReports/DiffModal/DiffModalActions.js:
```diff
@@ -0,0 +1,15 @@
+    type: DIFF_MODAL_CREATE,
+    payload: {
+      patch,
+      log,
```
please rename to title/filename?

glekner commented:
Included a GIF to the post

ohadlevy commented on webpack/assets/javascripts/react_app/components/ConfigReports/DiffModal/DiffModalActions.js:
```diff
@@ -0,0 +1,15 @@
+  dispatch({
+    type: DIFF_MODAL_CREATE,
+    payload: {
+      patch,
```
perhaps call it diff across ? while they both mean the same thing we do call the component diff and not patch.

ohadlevy commented on webpack/assets/javascripts/react_app/components/ConfigReports/DiffModal/DiffModalReducer.js:
```diff
@@ -0,0 +1,23 @@
+      return state.set('isOpen', !state.isOpen);
+    case DIFF_MODAL_CREATE:
+      return state
+        .set('patch', action.payload.patch)
```
please use merge.

glekner commented:
@ohadlevy applied suggestions.

ohadlevy commented on app/helpers/reports_helper.rb:
```diff
@@ -67,4 +67,11 @@ def report_origin_output_partial(origin)
+
+  def config_report_content(message, title)
+    if message.start_with?("\n---")
+      return link_to(_('Show Diff'), '#', data: {patch: message, log: title}, onclick: 'tfm.configReportsModalDiff.showDiff(this);')
```
why keep the log name here? imho:
```rb
def config_report_content(log)
    message = log.message.to_s
    filename = log.title.to_s # TODO extract the actual filename from the string
￼     if message.start_with?("\n---")
￼       return link_to(_('Show Diff'), '#', data: {patch: message, title: filename}, onclick: 'tfm.configReportsModalDiff.showDiff(this);')
    end
    message
  end
end

ohadlevy commented on app/views/config_reports/_output.html.erb:
```diff
@@ -11,7 +11,7 @@
         <td><span <%= report_tag log.level %>><%= h log.level %></span></td>
         <td class="break-me"><%= h log.source %></td>
-        <td class="break-me"><%= h log.message %></td>
+        <td class="break-me"><%= config_report_content(log.message.to_s, log.source.value) %></td>
```
based on the above comment, change this line to:
```erb
<td class="break-me"><%= config_report_content(log) %></td>
```

ohadlevy commented on webpack/assets/javascripts/foreman_config_reports_modal_diff.js:
```diff
@@ -0,0 +1,7 @@
+import * as diffModalActions from './react_app/components/ConfigReports/DiffModal/DiffModalActions';
+
+export const showDiff = (patch) => {
+  store.dispatch(diffModalActions.createDiff(patch.dataset.patch, patch.dataset.log));
```
should change log to diff and  title?

ohadlevy commented on webpack/assets/javascripts/foreman_config_reports_modal_diff.js:
```diff
@@ -0,0 +1,7 @@
+
+import * as diffModalActions from './react_app/components/ConfigReports/DiffModal/DiffModalActions';
+
+export const showDiff = (patch) => {
```
maybe it should be log vs patch? so its log.title and log.diff?

ohadlevy commented on webpack/assets/javascripts/react_app/components/ConfigReports/DiffModal/__tests__/DiffModalReducer.test.js:
```diff
@@ -0,0 +1,26 @@
+    action: {
+      type: DIFF_MODAL_CREATE,
+      payload: {
+        patch: diffModalMock.patch,
```
diff / title?

ohadlevy commented on webpack/assets/javascripts/react_app/components/DiffView/DiffView.js:
```diff
@@ -14,24 +14,47 @@ const getDiff = (oldText, newText) => {
+
+    return hunk && <Diff hunks={hunk} markEdits={markEdits} viewType={viewType} />;
+  }
+  // Patch
```
should we check we actually have a patch?

glekner replied:
i don't think the component should even be rendered in this case. I think its the responsibility of the parent.

ohadlevy commented on webpack/assets/javascripts/react_app/components/DiffView/DiffView.js:
```diff
@@ -14,24 +14,47 @@ const getDiff = (oldText, newText) => {
+  oldText: PropTypes.string,
+  newText: PropTypes.string,
   viewType: PropTypes.string.isRequired,
+  patch: PropTypes.string,
```
can you describe that on of the two are required(old/new or patch)?

ohadlevy commented on webpack/assets/javascripts/react_app/components/componentRegistry.js:
```diff
@@ -18,6 +18,8 @@ import EmptyState from './common/EmptyState';
 import ChartBox from './statistics/ChartBox';
 import ConfigReports from './ConfigReports/ConfigReports';
+import DiffContainer from './DiffView/DiffContainer';
+import DiffModal from './ConfigReports/DiffModal';
```
do we need both?

ohadlevy commented on app/helpers/reports_helper.rb:
```diff
@@ -67,4 +67,13 @@ def report_origin_output_partial(origin)
+
+  def config_report_content(log)
+    message = log.message.to_s
+    filename = log.source.value.to_s.scan(/\[(.*?)\]/).flatten.first
```
perhaps be defensive here, e.g.
```rb
filename = log.source.value.to_s.scan(/\[(.*?)\]/).flatten.first rescue ""
```

ohadlevy commented on webpack/assets/javascripts/react_app/components/ConfigReports/DiffModal/__tests__/DiffModalActions.test.js:
```diff
@@ -0,0 +1,12 @@
+const fixtures = {
+  'should toggleModal': () => toggleModal(),
+
+  'should createDiff': () => createDiff(diffModalMock.patch, 'log'),
```
should this be diff/title?

ohadlevy commented:
@theforeman/ui-ux can you please review? LGTM

ohadlevy commented on app/helpers/reports_helper.rb:
```diff
@@ -67,4 +67,13 @@ def report_origin_output_partial(origin)
+
+  def config_report_content(log)
+    message = log.message.to_s
+    filename = log.source.value.to_s.scan(/\[(.*?)\]/).flatten.first rescue ""
```
probably makes sense to move the filename inside the if condition.

ohadlevy commented on app/views/config_reports/show.html.erb:
```diff
@@ -1,4 +1,3 @@
-<%= javascript 'reports', 'charts' %>
```
nice, another page that wont need turbolinks full reload.

ohadlevy commented on app/views/config_reports/show.html.erb:
```diff
@@ -20,20 +19,6 @@
   link_to(_("Host details"), @config_report.host, :class => 'btn btn-default'),
   link_to(_("Other reports for this host"), host_config_reports_path(@config_report.host), :class => 'btn btn-default')
 %>
+  <div id="diff-modal-editor" class="modal-body"></div>
```
nitpick: extrawhitespaces before the the diff, should be with no leading spaces.

ohadlevy commented:
@glekner could you please also add missing tests (as seen in https://coveralls.io/builds/20204947)?

amirfefer commented on webpack/assets/javascripts/foreman_config_reports_modal_diff.js:
```diff
@@ -0,0 +1,7 @@
+
+import * as diffModalActions from './react_app/components/ConfigReports/DiffModal/DiffModalActions';
+
+export const showDiff = (log) => {
```
can you add a test for it?

glekner replied:
as discussed with @sharvit before, we decided not to test files that are only responsible for `store.dispatch()` funcs

amirfefer commented on webpack/assets/javascripts/react_app/components/ConfigReports/DiffModal/DiffModal.js:
```diff
@@ -0,0 +1,49 @@
+import DiffContainer from '../../DiffView/DiffContainer';
+import './diffmodal.scss';
+
+class DiffModal extends React.Component {
```
can you switch this to a functional component?

glekner commented:
@ohadlevy `foreman_config_reports_modal_diff.js` only has a `store.dispatch()` function, just like `foreman_navigation.js`, we agreed to not test these files as the actions are being tested in their own file.

glekner commented:
- Moved the diffViewType buttons to the Modal header for better space usage
<img width="917" alt="screenshot 2018-11-21 14 17 23" src="https://user-images.githubusercontent.com/24938324/48840788-557f0600-ed98-11e8-9117-83894b1e0898.png">

- Added `text-overflow: ellipsis` to the title in case of long filenames
<img width="338" alt="screenshot 2018-11-21 14 10 12" src="https://user-images.githubusercontent.com/24938324/48840844-70ea1100-ed98-11e8-9488-e7e0057dd4da.png">

ohadlevy commented:
@glekner can you please add stroybook?

amirfefer commented on webpack/assets/javascripts/react_app/components/ConfigReports/DiffModal/DiffModal.js:
```diff
@@ -0,0 +1,54 @@
+      title, diff, isOpen, toggleModal, diffViewType, changeViewType,
+    } = this.props;
+    return (
+      <Modal show={isOpen} onHide={() => toggleModal()} className="diff-modal">
```
how about passing toggleModal directly to `onHide`, instead of another anonymous function? 
 `onHide={toggleModal}`

ohadlevy commented:
@amirfefer / @sharvit can you guys review please? thanks.

amirfefer reviewed (changes requested):
Thanks @glekner, looks great :+1: 
left few minor issues
